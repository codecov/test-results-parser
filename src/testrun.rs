use pyo3::class::basic::CompareOp;
use pyo3::{prelude::*, pyclass};

use crate::helpers::s;

#[derive(Clone, Debug)]
#[pyclass]
pub struct Testrun {
    pub name: String,
    pub duration: String,
    pub outcome: String,
    pub testsuite: String,
}

impl Testrun {
    pub fn empty() -> Testrun {
        Testrun {
            name: s(""),
            duration: s(""),
            outcome: s(""),
            testsuite: s(""),
        }
    }
}

#[pymethods]
impl Testrun {
    #[new]
    fn new(name: String, duration: String, outcome: String, testsuite: String) -> Self {
        Self {
            name,
            duration,
            outcome,
            testsuite,
        }
    }

    fn __repr__(&self) -> String {
        format!(
            "({}, {}, {}, {})",
            self.name, self.outcome, self.duration, self.testsuite
        )
    }

    fn __richcmp__(&self, other: &Self, op: CompareOp) -> PyResult<bool> {
        match op {
            CompareOp::Eq => Ok(self.name == other.name
                && self.outcome == other.outcome
                && self.duration == other.duration
                && self.testsuite == other.testsuite),
            _ => todo!(),
        }
    }
}
