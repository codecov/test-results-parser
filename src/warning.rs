use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct WarningInfo {
    pub message: String,
    pub location: u64,
}

impl WarningInfo {
    pub fn new(message: String, location: u64) -> Self {
        Self { message, location }
    }
}
