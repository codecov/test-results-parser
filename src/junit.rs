use anyhow::{Context, Result};
use pyo3::prelude::*;
use std::collections::HashSet;

use quick_xml::events::attributes::{Attribute, Attributes};
use quick_xml::events::{BytesStart, Event};
use quick_xml::reader::Reader;

use crate::compute_name::{compute_name, unescape_str};
use crate::testrun::{check_testsuites_name, Framework, Outcome, Testrun};
use crate::validated_string::ValidatedString;
use crate::warning::WarningInfo;

fn convert_attribute(attribute: Attribute) -> Result<String> {
    let bytes = attribute.value.into_owned();
    let value = String::from_utf8(bytes).context("Error converting attribute to string")?;
    Ok(value)
}
struct TestcaseAttrs {
    name: ValidatedString,
    time: Option<String>,
    classname: Option<ValidatedString>,
    file: Option<ValidatedString>,
}

enum AttrsOrWarning {
    Attributes(TestcaseAttrs),
    Warning(String),
}

// originally from https://gist.github.com/scott-codecov/311c174ecc7de87f7d7c50371c6ef927#file-cobertura-rs-L18-L31
fn parse_testcase_attrs(attributes: Attributes) -> Result<AttrsOrWarning> {
    let mut name: Option<ValidatedString> = None;
    let mut time: Option<String> = None;
    let mut classname: Option<ValidatedString> = None;
    let mut file: Option<ValidatedString> = None;

    for attribute in attributes {
        let attribute = attribute.context("Error parsing attribute")?;

        match attribute.key.into_inner() {
            b"time" => {
                time = Some(convert_attribute(attribute)?);
            }
            b"classname" => {
                let unvalidated_classname = convert_attribute(attribute)?;
                classname = match ValidatedString::from_string(unvalidated_classname) {
                    Ok(name) => Some(name),
                    Err(_) => {
                        return Ok(AttrsOrWarning::Warning("Error validating classname".into()));
                    }
                };
            }
            b"name" => {
                let unvalidated_name = convert_attribute(attribute)?;
                name = match ValidatedString::from_string(unvalidated_name) {
                    Ok(name) => Some(name),
                    Err(_) => {
                        return Ok(AttrsOrWarning::Warning("Error validating name".into()));
                    }
                };
            }
            b"file" => {
                let unvalidated_file = convert_attribute(attribute)?;
                file = match ValidatedString::from_string(unvalidated_file) {
                    Ok(name) => Some(name),
                    Err(_) => {
                        return Ok(AttrsOrWarning::Warning("Error validating file".into()));
                    }
                };
            }
            _ => {}
        }
    }

    match name {
        Some(name) => Ok(AttrsOrWarning::Attributes(TestcaseAttrs {
            name,
            time,
            classname,
            file,
        })),
        None => anyhow::bail!("No name found"),
    }
}

fn get_attribute(e: &BytesStart, name: &str) -> Result<Option<String>> {
    let attr = if let Some(message) = e
        .try_get_attribute(name)
        .context("Error parsing attribute")?
    {
        Some(String::from_utf8(message.value.to_vec())?)
    } else {
        None
    };
    Ok(attr)
}

fn populate(
    rel_attrs: TestcaseAttrs,
    testsuite: ValidatedString,
    testsuite_time: Option<&str>,
    framework: Option<Framework>,
    network: Option<&HashSet<String>>,
) -> Result<(Testrun, Option<Framework>)> {
    let name = rel_attrs.name;
    let classname = rel_attrs.classname.unwrap_or_default();
    let duration = rel_attrs
        .time
        .as_deref()
        .or(testsuite_time)
        .and_then(|t| t.parse().ok());
    let file = rel_attrs.file;

    let mut t = Testrun {
        name,
        classname,
        duration,
        outcome: Outcome::Pass,
        testsuite,
        failure_message: None,
        filename: file,
        build_url: None,
        computed_name: ValidatedString::default(),
    };

    let framework = framework.or_else(|| t.framework());
    let computed_name = compute_name(
        &t.classname,
        &t.name,
        framework,
        t.filename.as_deref(),
        network,
    );
    t.computed_name = ValidatedString::from_string(computed_name)
        .context("Error converting computed name to ValidatedString")?;

    Ok((t, framework))
}

pub fn get_position_info(input: &[u8], byte_offset: usize) -> (usize, usize) {
    let mut line = 1;
    let mut last_newline = 0;

    for (i, &byte) in input.iter().take(byte_offset).enumerate() {
        if byte == b'\n' {
            line += 1;
            last_newline = i + 1;
        }
    }

    let column = byte_offset - last_newline + 1;

    (line, column)
}

enum TestrunOrSkipped {
    Testrun(Testrun),
    Skipped,
}

pub fn use_reader(
    reader: &mut Reader<&[u8]>,
    network: Option<&HashSet<String>>,
) -> PyResult<(Option<Framework>, Vec<Testrun>, Vec<WarningInfo>)> {
    let mut testruns: Vec<Testrun> = Vec::new();
    let mut saved_testrun: Option<TestrunOrSkipped> = None;

    let mut in_failure: bool = false;
    let mut in_error: bool = false;

    let mut framework: Option<Framework> = None;

    let mut warnings: Vec<WarningInfo> = Vec::new();

    // every time we come across a testsuite element we update this vector:
    // if the testsuite element contains the time attribute append its value to this vec
    // else append a clone of the last value in the vec
    let mut testsuite_names: Vec<Option<ValidatedString>> = vec![];
    let mut testsuite_times: Vec<Option<String>> = vec![];

    let mut buf = Vec::new();
    loop {
        let event = reader
            .read_event_into(&mut buf)
            .context("Error parsing XML")?;
        match event {
            Event::Eof => {
                break;
            }
            Event::Start(e) => match e.name().as_ref() {
                b"testcase" => {
                    let attrs = parse_testcase_attrs(e.attributes())?;
                    match attrs {
                        AttrsOrWarning::Attributes(attrs) => {
                            let (testrun, parsed_framework) = populate(
                                attrs,
                                testsuite_names
                                    .iter()
                                    .rev()
                                    .find_map(|e| e.clone())
                                    .unwrap_or_default(),
                                testsuite_times.iter().rev().find_map(|e| e.as_deref()),
                                framework,
                                network,
                            )?;
                            saved_testrun = Some(TestrunOrSkipped::Testrun(testrun));
                            framework = parsed_framework;
                        }
                        AttrsOrWarning::Warning(warning) => {
                            warnings.push(WarningInfo::new(warning, reader.buffer_position()));
                            saved_testrun = Some(TestrunOrSkipped::Skipped);
                        }
                    }
                }
                b"skipped" => {
                    let saved = saved_testrun
                        .as_mut()
                        .context("Error accessing saved testrun")?;
                    match saved {
                        TestrunOrSkipped::Testrun(testrun) => {
                            testrun.outcome = Outcome::Skip;
                        }
                        TestrunOrSkipped::Skipped => {}
                    }
                }
                b"error" => {
                    let saved = saved_testrun
                        .as_mut()
                        .context("Error accessing saved testrun")?;
                    match saved {
                        TestrunOrSkipped::Testrun(testrun) => {
                            testrun.outcome = Outcome::Error;

                            testrun.failure_message = get_attribute(&e, "message")?
                                .map(|failure_message| unescape_str(&failure_message).into());
                        }
                        TestrunOrSkipped::Skipped => {}
                    }

                    in_error = true;
                }
                b"failure" => {
                    let saved = saved_testrun
                        .as_mut()
                        .context("Error accessing saved testrun")?;
                    match saved {
                        TestrunOrSkipped::Testrun(testrun) => {
                            testrun.outcome = Outcome::Failure;

                            testrun.failure_message = get_attribute(&e, "message")?
                                .map(|failure_message| unescape_str(&failure_message).into());
                        }
                        TestrunOrSkipped::Skipped => {}
                    }

                    in_failure = true;
                }
                b"testsuite" => {
                    testsuite_names.push(
                        get_attribute(&e, "name")?
                            .map(|s| {
                                ValidatedString::from_string(s)
                                    .context("Error converting testsuite name to ValidatedString")
                            })
                            .transpose()?,
                    );
                    testsuite_times.push(get_attribute(&e, "time")?);
                }
                b"testsuites" => {
                    let testsuites_name = get_attribute(&e, "name")?;
                    framework = testsuites_name.and_then(|name| check_testsuites_name(&name))
                }
                _ => {}
            },
            Event::End(e) => match e.name().as_ref() {
                b"testcase" => {
                    let saved = saved_testrun.take().context(
                        "Met testcase closing tag without first meeting testcase opening tag",
                    )?;
                    match saved {
                        TestrunOrSkipped::Testrun(testrun) => testruns.push(testrun),
                        TestrunOrSkipped::Skipped => {}
                    }
                }
                b"failure" => in_failure = false,
                b"error" => in_error = false,
                b"testsuite" => {
                    testsuite_times.pop();
                    testsuite_names.pop();
                }
                _ => (),
            },
            Event::Empty(e) => match e.name().as_ref() {
                b"testcase" => {
                    let rel_attrs = parse_testcase_attrs(e.attributes())?;
                    match rel_attrs {
                        AttrsOrWarning::Attributes(attrs) => {
                            let (testrun, parsed_framework) = populate(
                                attrs,
                                testsuite_names
                                    .iter()
                                    .rev()
                                    .find_map(|e| e.clone())
                                    .unwrap_or_default(),
                                testsuite_times.iter().rev().find_map(|e| e.as_deref()),
                                framework,
                                network,
                            )?;
                            testruns.push(testrun);
                            framework = parsed_framework;
                        }
                        AttrsOrWarning::Warning(warning) => {
                            warnings.push(WarningInfo::new(warning, reader.buffer_position()));
                        }
                    }
                }
                b"failure" => {
                    let saved = saved_testrun
                        .as_mut()
                        .context("Error accessing saved testrun")?;
                    match saved {
                        TestrunOrSkipped::Testrun(testrun) => {
                            testrun.outcome = Outcome::Failure;

                            testrun.failure_message = get_attribute(&e, "message")?
                                .map(|failure_message| unescape_str(&failure_message).into());
                        }
                        TestrunOrSkipped::Skipped => {}
                    }
                }
                b"skipped" => {
                    let saved = saved_testrun
                        .as_mut()
                        .context("Error accessing saved testrun")?;
                    match saved {
                        TestrunOrSkipped::Testrun(testrun) => {
                            testrun.outcome = Outcome::Skip;
                        }
                        TestrunOrSkipped::Skipped => {}
                    }
                }
                b"error" => {
                    let saved = saved_testrun
                        .as_mut()
                        .context("Error accessing saved testrun")?;
                    match saved {
                        TestrunOrSkipped::Testrun(testrun) => {
                            testrun.outcome = Outcome::Error;

                            testrun.failure_message = get_attribute(&e, "message")?
                                .map(|failure_message| unescape_str(&failure_message).into());
                        }
                        TestrunOrSkipped::Skipped => {}
                    }
                }
                _ => {}
            },
            Event::Text(mut xml_failure_message) => {
                if in_failure || in_error {
                    let saved = saved_testrun
                        .as_mut()
                        .context("Error accessing saved testrun")?;
                    match saved {
                        TestrunOrSkipped::Testrun(testrun) => {
                            xml_failure_message.inplace_trim_end();
                            xml_failure_message.inplace_trim_start();

                            testrun.failure_message = Some(
                                unescape_str(std::str::from_utf8(&xml_failure_message)?).into(),
                            );
                        }
                        TestrunOrSkipped::Skipped => {}
                    }
                }
            }

            // There are several other `Event`s we do not consider here
            _ => (),
        }
        buf.clear()
    }

    Ok((framework, testruns, warnings))
}
