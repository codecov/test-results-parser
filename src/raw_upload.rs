use anyhow::Context;

use base64::prelude::*;
use pyo3::prelude::*;
use std::collections::HashSet;
use std::io::prelude::*;

use flate2::bufread::ZlibDecoder;

use quick_xml::reader::Reader;
use serde::Deserialize;

use crate::junit::{get_position_info, use_reader};
use crate::testrun::ParsingInfo;
use crate::warning::WarningInfo;

#[derive(Deserialize, Debug, Clone)]
struct TestResultFile {
    filename: String,
    data: String,
}
#[derive(Deserialize, Debug, Clone)]
struct RawTestResultUpload {
    #[serde(default)]
    network: Option<HashSet<String>>,
    test_results_files: Vec<TestResultFile>,
}

#[derive(Debug, Clone)]
struct ReadableFile {
    filename: String,
    data: Vec<u8>,
}

const LEGACY_FORMAT_PREFIX: &[u8] = b"# path=";
const LEGACY_FORMAT_SUFFIX: &[u8] = b"<<<<<< EOF";

fn serialize_to_legacy_format(readable_files: Vec<ReadableFile>) -> Vec<u8> {
    let mut res = Vec::new();
    for file in readable_files {
        res.extend_from_slice(LEGACY_FORMAT_PREFIX);
        res.extend_from_slice(file.filename.as_bytes());
        res.extend_from_slice(b"\n");
        res.extend_from_slice(&file.data);
        res.extend_from_slice(b"\n");
        res.extend_from_slice(LEGACY_FORMAT_SUFFIX);
        res.extend_from_slice(b"\n");
    }
    res
}

// the warnings should be ordered by location because they're pushed to the vec as we parse
// so we can guarantee that warning[x].location >= warning[x - 1].location
// implicitly tested by warnings-junit.xml
fn format_warnings(input: &[u8], warnings: Vec<WarningInfo>, filename: &str) -> Vec<String> {
    let mut offset = 0;
    let mut result = Vec::new();
    let mut line = 1;
    let mut col = 0;
    let mut input_iter = input.iter();
    for warning in warnings {
        for bytes in input_iter
            .by_ref()
            .take((warning.location - offset) as usize)
        {
            if *bytes == b'\n' {
                line += 1;
                col = 0;
            } else {
                col += 1;
            }
        }
        offset += warning.location;
        result.push(if warning.message.contains('\n') {
            format!("{}\n\nAt {}:{} in {}", warning.message, line, col, filename)
        } else {
            format!("{} at {}:{} in {}", warning.message, line, col, filename)
        });
    }
    result
}

#[pyfunction]
#[pyo3(signature = (raw_upload_bytes))]
pub fn parse_raw_upload(raw_upload_bytes: &[u8]) -> anyhow::Result<(Vec<ParsingInfo>, Vec<u8>)> {
    let upload: RawTestResultUpload =
        serde_json::from_slice(raw_upload_bytes).context("Error deserializing json")?;
    let network: Option<HashSet<String>> = upload.network;

    let mut results: Vec<ParsingInfo> = Vec::with_capacity(upload.test_results_files.len());
    let mut readable_files: Vec<ReadableFile> = Vec::with_capacity(upload.test_results_files.len());

    for file in upload.test_results_files {
        let decoded_file_bytes = BASE64_STANDARD
            .decode(file.data)
            .context("Error decoding base64")?;

        let mut decoder = ZlibDecoder::new(decoded_file_bytes.as_slice());

        let mut decompressed_file_bytes = Vec::new();
        decoder
            .read_to_end(&mut decompressed_file_bytes)
            .context("Error decompressing file")?;

        let mut reader = Reader::from_reader(decompressed_file_bytes.as_slice());
        reader.config_mut().trim_text(true);
        let (framework, testruns, warnings) = use_reader(&mut reader, network.as_ref())
            .with_context(|| {
                let pos_conversion = reader.buffer_position().try_into();
                match pos_conversion {
                    Ok(pos) => {
                        let (line, col) = get_position_info(&decompressed_file_bytes, pos);
                        format!(
                            "Error parsing JUnit XML in {} at {}:{}",
                            file.filename, line, col
                        )
                    }
                    Err(_) => format!("Error parsing JUnit XML in {}", file.filename),
                }
            })?;

        let warning_strings: Vec<String> =
            format_warnings(&decompressed_file_bytes, warnings, &file.filename);

        let parsing_info = ParsingInfo {
            framework,
            testruns,
            warnings: warning_strings,
        };
        results.push(parsing_info);

        let readable_file = ReadableFile {
            data: decompressed_file_bytes,
            filename: file.filename,
        };
        readable_files.push(readable_file);
    }

    let readable_file = serialize_to_legacy_format(readable_files);

    Ok((results, readable_file))
}

#[cfg(test)]
mod tests {
    use base64::engine::general_purpose::STANDARD as BASE64_STANDARD;
    use base64::Engine;
    use flate2::Compression;
    use std::io::Write;

    use super::*;
    use insta::{assert_yaml_snapshot, glob};

    fn file_into_bytes(filename: &str) -> Vec<u8> {
        let upload = std::fs::read(filename).unwrap();
        let mut encoder = flate2::write::ZlibEncoder::new(Vec::new(), Compression::default());
        encoder.write_all(&upload).unwrap();
        let compressed = encoder.finish().unwrap();
        let base64_data = BASE64_STANDARD.encode(compressed);
        let upload_json = format!(
            r#"{{"network": [], "test_results_files": [{{"filename": "{}", "format": "base64+compressed", "data": "{}"}}]}}"#,
            filename.split('/').next_back().unwrap(),
            base64_data,
        );
        upload_json.into()
    }

    #[test]
    fn test_parse_raw_upload_success() {
        glob!("../tests", "*.xml", |path| {
            let upload_json = file_into_bytes(path.to_str().unwrap());
            let result = parse_raw_upload(&upload_json);
            match result {
                Ok((results, _)) => assert_yaml_snapshot!(results),
                Err(e) => {
                    assert_yaml_snapshot!(e.to_string());
                }
            }
        });
    }
}
