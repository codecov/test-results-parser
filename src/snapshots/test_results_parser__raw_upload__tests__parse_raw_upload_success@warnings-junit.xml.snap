---
source: src/raw_upload.rs
expression: results
input_file: tests/warnings-junit.xml
---
- framework: Jest
  testruns: []
  warnings:
    - "Error validating classname ending at 7:65 in warnings-junit.xml"
    - "Error validating classname ending at 11:65 in warnings-junit.xml"
